### PR TITLE
thread-per-simulation parallelization

### PR DESCRIPTION
**Pull request by ransomw1c**
spin off a thread for every simulation, doing the simplest thing first.

an alternative is to use a [fixed number of threads](https://clojure.org/about/concurrent_programming) and distribute the simulations uniformly among them.

`core.async`, the Clojure version of CSP (goroutines), is not used because it doesn't necessarily map onto actual parallelism across cores in the same way that CSP does in golang.  using the JVM mechanisms directly is the recommended way to ensure actual parallelism in Clojure.

### CODE REVIEW

**Comment by mattsdni** [comment]
out of curiosity, have you measured the performance impact of this change?

**Comment by ransomw1c** [comment]
i have not.  i could add [`profiling`](https://github.com/ptaoussanis/tufte)..

File: dev/user.clj [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+         "\n"
+         [(str "CREATE ROLE " gridfire-user " WITH LOGIN CREATEDB;")
+          (str "CREATE DATABASE " gridfire-db-name " WITH OWNER " gridfire-user ";")
+          (str "\\c " gridfire-db-name)
+          "CREATE EXTENSION postgis;"])]
+    (conch/with-programs [raster2pgsql psql createuser]
+      (createuser "-s" pg-user)
+      (psql "-U" pg-user {:in (to-rdr init-gridfire-sql)})
+      (psql "-U" pg-user
+            {:in (raster2pgsql "-s" "4326"
```
Comment (itsalitee) [inline dev/user.clj]: Could the `srid` , i.e. `4326`, be a variable here?

File: dev/user.clj [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                               "test/input_data/tubbs_1507528800_1507530600.tif"
+                               "dem")})
+      (doseq [[fuel-tag-idx fuel-tag]
+              (map list (map inc (range))
+                   fuel-tags)]
+        (psql "-U" gridfire-user "gridfire"
+              {:in (raster2pgsql "-s" "4326"
```
Comment (itsalitee) [inline dev/user.clj]: Same as the previous question.

File: src/gridfire/cli.clj [hunk 3]
```diff
@@ -135,68 +136,78 @@
...
-               :max-runtime               (max-runtime i)
-               :temperature               (temperature i)
-               :relative-humidity         (relative-humidity i)
-               :wind-speed-20ft           (wind-speed-20ft i)
-               :wind-from-direction       (wind-from-direction i)
-               :foliar-moisture           (foliar-moisture i)
-               :ellipse-adjustment-factor (ellipse-adjustment-factor i)}
-              (summarize-fire-spread-results fire-spread-results cell-size))))
+   outfile-suffix output-geotiffs? output-pngs? output-csvs?
+   i]
```
Comment (itsalitee) [inline src/gridfire/cli.clj:146]: Would really help if we describe the logic for removing the `map`.

**Review by itsalitee — COMMENTED** [review]
Thanks a lot @ransomw1c for your efforts on this work! I had minor comments/suggestions.

**Comment by ransomw1c** [comment]
profiling results show a marked speedup on desktop.  here's the result of sequential simulations

```
19-08-27 20:13:44 UnknownHost DEBUG [gridfire.cli:212] - running simulations
19-08-27 20:13:44 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:13:49 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:13:53 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:13:53 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:13:58 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:14:03 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:14:08 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:14:12 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:14:22 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:14:27 UnknownHost DEBUG [gridfire.cli:147] - running simulation

                          pId     nCalls        Min      50% ≤      90% ≤      95% ≤      99% ≤        Max       Mean   MAD       Total   Clock

 :gridfire.cli/run-simulation         10    20.76μs     4.67s      4.77s      4.99s      4.99s      4.99s      4.22s   ±20%     42.22s      75%
:gridfire.cli/run-simulations          1   340.68μs   340.68μs   340.68μs   340.68μs   340.68μs   340.68μs   340.68μs   ±0%    340.68μs      0%

                    Accounted                                                                                                   42.22s      75%
                        Clock                                                                                                   56.26s     100%
```

with each simulation taking 4 or 5 seconds and all 10 taking about a minute.

here's the result of parallelized simulations

```
19-08-27 20:16:44 UnknownHost DEBUG [gridfire.cli:212] - running simulations
19-08-27 20:16:49 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:16:49 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:16:49 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:16:49 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:16:49 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:16:49 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:16:49 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:16:49 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:16:49 UnknownHost INFO [gridfire.cli:221] -                          pId     nCalls        Min      50% ≤      90% ≤      95% ≤      99% ≤        Max       Mean   MAD       Total   Clock

:gridfire.cli/run-simulation          1    24.72μs    24.72μs    24.72μs    24.72μs    24.72μs    24.72μs    24.72μs   ±0%     24.72μs      1%

                   Accounted                                                                                                   24.72μs      1%
                       Clock                                                                                                    1.97ms    100%

19-08-27 20:16:49 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:16:49 UnknownHost DEBUG [gridfire.cli:147] - running simulation
19-08-27 20:16:59 UnknownHost INFO [gridfire.cli:221] -                          pId     nCalls        Min      50% ≤      90% ≤      95% ≤      99% ≤        Max       Mean   MAD       Total   Clock

:gridfire.cli/run-simulation          1     9.57s      9.57s      9.57s      9.57s      9.57s      9.57s      9.57s    ±0%      9.57s     100%

                   Accounted                                                                                                    9.57s     100%
                       Clock                                                                                                    9.57s     100%

19-08-27 20:16:59 UnknownHost INFO [gridfire.cli:221] -                          pId     nCalls        Min      50% ≤      90% ≤      95% ≤      99% ≤        Max       Mean   MAD       Total   Clock

:gridfire.cli/run-simulation          1     9.60s      9.60s      9.60s      9.60s      9.60s      9.60s      9.60s    ±0%      9.60s     100%

                   Accounted                                                                                                    9.60s     100%
                       Clock                                                                                                    9.60s     100%

19-08-27 20:16:59 UnknownHost INFO [gridfire.cli:221] -                          pId     nCalls        Min      50% ≤      90% ≤      95% ≤      99% ≤        Max       Mean   MAD       Total   Clock

:gridfire.cli/run-simulation          1     9.64s      9.64s      9.64s      9.64s      9.64s      9.64s      9.64s    ±0%      9.64s     100%

                   Accounted                                                                                                    9.64s     100%
                       Clock                                                                                                    9.64s     100%

19-08-27 20:16:59 UnknownHost INFO [gridfire.cli:221] -                          pId     nCalls        Min      50% ≤      90% ≤      95% ≤      99% ≤        Max       Mean   MAD       Total   Clock

:gridfire.cli/run-simulation          1     9.70s      9.70s      9.70s      9.70s      9.70s      9.70s      9.70s    ±0%      9.70s     100%

                   Accounted                                                                                                    9.70s     100%
                       Clock                                                                                                    9.71s     100%

19-08-27 20:16:59 UnknownHost INFO [gridfire.cli:221] -                          pId     nCalls        Min      50% ≤      90% ≤      95% ≤      99% ≤        Max       Mean   MAD       Total   Clock

:gridfire.cli/run-simulation          1     9.72s      9.72s      9.72s      9.72s      9.72s      9.72s      9.72s    ±0%      9.72s     100%

                   Accounted                                                                                                    9.72s     100%
                       Clock                                                                                                    9.73s     100%

19-08-27 20:16:59 UnknownHost INFO [gridfire.cli:221] -                          pId     nCalls        Min      50% ≤      90% ≤      95% ≤      99% ≤        Max       Mean   MAD       Total   Clock

:gridfire.cli/run-simulation          1     9.78s      9.78s      9.78s      9.78s      9.78s      9.78s      9.78s    ±0%      9.78s     100%

                   Accounted                                                                                                    9.78s     100%
                       Clock                                                                                                    9.79s     100%

19-08-27 20:16:59 UnknownHost INFO [gridfire.cli:221] -                          pId     nCalls        Min      50% ≤      90% ≤      95% ≤      99% ≤        Max       Mean   MAD       Total   Clock

:gridfire.cli/run-simulation          1     9.80s      9.80s      9.80s      9.80s      9.80s      9.80s      9.80s    ±0%      9.80s     100%

                   Accounted                                                                                                    9.80s     100%
                       Clock                                                                                                    9.80s     100%

19-08-27 20:16:59 UnknownHost INFO [gridfire.cli:221] -                          pId     nCalls        Min      50% ≤      90% ≤      95% ≤      99% ≤        Max       Mean   MAD       Total   Clock

:gridfire.cli/run-simulation          1     9.84s      9.84s      9.84s      9.84s      9.84s      9.84s      9.84s    ±0%      9.84s     100%

                   Accounted                                                                                                    9.84s     100%
                       Clock                                                                                                    9.84s     100%

19-08-27 20:16:59 UnknownHost INFO [gridfire.cli:221] -                          pId     nCalls        Min      50% ≤      90% ≤      95% ≤      99% ≤        Max       Mean   MAD       Total   Clock

:gridfire.cli/run-simulation          1    10.22s     10.22s     10.22s     10.22s     10.22s     10.22s     10.22s    ±0%     10.22s     100%

                   Accounted                                                                                                   10.22s     100%
                       Clock                                                                                                   10.22s     100%


                          pId     nCalls        Min      50% ≤      90% ≤      95% ≤      99% ≤        Max       Mean   MAD       Total   Clock

:gridfire.cli/run-simulations          1    10.22s     10.22s     10.22s     10.22s     10.22s     10.22s     10.22s    ±0%     10.22s      42%

                    Accounted                                                                                                   10.22s      42%
                        Clock                                                                                                   24.49s     100%
```

with each simulation taking 9 or 10 seconds such that all simulations combined take about 10 seconds.

File: dev/user.clj (hunk 1 above)
Comment (ransomw1c) [reply dev/user.clj]: i've parameterized srid.  here and in `startup-health-check`, passing `:srid "<srid>"` as a "vararg" (as with `:db-username`, etc. in the `comment`) will change the value passed in.

i didn't parameterize it initially b/c it was unclear to me whether this is something that can be changed independently of the rest of the codebase.  lack of postgis knowledge.

File: src/gridfire/cli.clj (hunk 3 above)
Comment (ransomw1c) [reply src/gridfire/cli.clj:146]: see if the intermediate commits here explain more

https://github.com/ransomw1c/gridfire/commits/jFIRE256-parallelization--wip-01

this is, specifically, where `mapv` of tasks over indexes is replaced with a `map` over [futures](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Future.html) (also known as promises).

https://github.com/ransomw1c/gridfire/commit/0a22a8c99a7a87400ce5c2e0831cb306349a927d

so the `map` isn't being removed:  there is still a map to extract the async results.  the synchronous s-exp has be separated out of the mapping/parallelization forms.

**Review by mattsdni — APPROVED** [review]
looks good. nice performance increases.

**Pull request merged**